src/quic/preferredaddress.h

```diff
@@ -0,0 +1,70 @@
+#pragma once
+
+#if defined(NODE_WANT_INTERNALS) && NODE_WANT_INTERNALS
+
+#include <env.h>
+#include <node_internals.h>
+#include <ngtcp2/ngtcp2.h>
+#include <string>
+#include <v8.h>
+
+namespace node {
+namespace quic {
+
+// PreferredAddress is a helper class used only when a client Session receives
+// an advertised preferred address from a server. The helper provides
+// information about the server advertised preferred address and allows
+// the preferred address to be selected.
+class PreferredAddress final {
+ public:
+  enum class Policy {
+    // Ignore the server-advertised preferred address.
+    IGNORE,
+    // Use the server-advertised preferred address.
+    USE,
+  };
+
+  // The QUIC_* constants are expected to be exported out to be used on
+  // the JavaScript side of the API.
+  static constexpr uint32_t QUIC_PREFERRED_ADDRESS_USE =
+      static_cast<uint32_t>(Policy::USE);
+  static constexpr uint32_t QUIC_PREFERRED_ADDRESS_IGNORE =
+      static_cast<uint32_t>(Policy::IGNORE);
+
+  static v8::Maybe<Policy> GetPolicy(Environment* env,
+                                     v8::Local<v8::Value> value);
+
+  struct AddressInfo final {
+    char host[NI_MAXHOST];
+    int family;
+    uint16_t port;
+    std::string_view address;
+  };
+
+  explicit PreferredAddress(ngtcp2_path* dest,
+                            const ngtcp2_preferred_addr* paddr);
+  PreferredAddress(const PreferredAddress&) = delete;
+  PreferredAddress(PreferredAddress&&) = delete;
+  PreferredAddress& operator=(const PreferredAddress&) = delete;
+  PreferredAddress& operator=(PreferredAddress&&) = delete;
+
+  void Use(const AddressInfo& address);
+
+  std::optional<const AddressInfo> ipv4() const;
+  std::optional<const AddressInfo> ipv6() const;
```

Comment:
Is it possible to avoid using `std::optional`? I remember @lemire's blog post about it: https://lemire.me/blog/2023/01/12/care-is-needed-to-use-c-stdoptional-with-non-trivial-objects/

Reply:
Citing that very blog post (emphasis added): "I **do not recommend against using std::optional**. There are effective strategies to avoid the copies. Just apply some care **if performance is a concern**."

Given QUIC's development status and the unbelievable amount of work that @jasnell is doing towards it already, I don't think this kind of (potential) performance optimization is a concern at this point, especially if it reduces simplicity, readability, or maintenability.

Reply:
Given that these are likely to be accessed at most once per client-side connection, I'm not convinced it's enough of an issue to worry about just yet. As @tniessen suggests, that's something we can optimize around later.